### PR TITLE
[inductor] Avoid Python fallback for cpp custom op lists

### LINKED ISSUE #153478: cpp wrapper calls back to python for custom op even when a C++ registration is made
### 🐛 Describe the bug

See issue: https://github.com/mingfeima/sglang/pull/73#issuecomment-2874712382, 

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @aakhundov @benjaminglass1 @desertfire 

### Versions

master

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184099

Use direct boxed dispatcher calls in cpp_wrapper JIT for runtime fallback custom ops whose schemas are outside the StableIValue bridge, such as Tensor[] arguments, while keeping the existing StableIValue path for supported schemas.

Fixes #153478

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by zou3519 — COMMENTED** [review]
@yushangdi @desertfire could you confirm that the code here does not affect the aoti path that does need stable abi?

**Comment by desertfire** [comment]
> @yushangdi @desertfire could you confirm that the code here does not affect the aoti path that does need stable abi?

Yes, this is fine. It doesn't touch the AOTI path.

**Review by desertfire — APPROVED** [review]

**Review by zou3519 — APPROVED** [review]
sg then

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main and resubmitted with ghstack. The mergebot Windows build failure hit the stale-base pip install issue that is fixed on current main by #185662; local validation after the rebase passed with `PYTHONPATH=. python test/inductor/test_cpp_wrapper_custom_ops.py TestCppWrapperCustomOps.test_cpp_custom_op_runtime_dispatch_stays_in_cpp` and `lintrunner -a`.

**Comment by jansel** [comment]
My agent says I fixed the internal test revert by making the inline C++ extension module name unique per test process, avoiding reuse of stale torch_extensions binaries on persistent CI workers while keeping the custom-op namespace stable within the process. Local validation passed with `PYTHONPATH=. python test/inductor/test_cpp_wrapper_custom_ops.py TestCppWrapperCustomOps.test_cpp_custom_op_runtime_dispatch_stays_in_cpp` and `lintrunner test/inductor/test_cpp_wrapper_custom_ops.py`; full `lintrunner -a` still hits an unrelated Pyrefly error in `torch/nn/modules/loss.py`.

**Comment by jansel** [comment]
Agent, below is an analysis of the revert.  Fix the issue if it hasn't been fixed already and explain what change.  How confident are you this is safe to reland?

# Revert Analysis: D107035517 — [inductor] Avoid Python fallback for cpp custom op lists

## Failing Tests

| # | Test | Time | Assessment |
| --- | --- | --- | --- |
| 1 | `fbcode//caffe2/test/inductor:cpp_wrapper_custom_ops` — `test_cpp_custom_op_runtime_dispatch_stays_in_cpp` | 0.074s | New test added by the diff; assertion failure |
| 2 | `fbcode//pytorch/ao/test:.../x86inductor_fusion` — `test_qconv2d_maxpool2d_linear_dynamic_cpu_cpp_wrapper` | 9.6s (4 retries) | Pre-existing test; exercises cpp_wrapper codegen for quantized ops |
| 3 | `fbcode//pytorch/ao/test:.../x86inductor_fusion` — `test_qconv2d_maxpool2d_linear_dynamic_cpu_dynamic_shapes_cpp_wrapper` | 23.4s (4 retries) | Dynamic-shapes variant of #2 |

**Note:** The PR's OSS CI on GitHub passed (HUD showed "✅ You can merge normally! (2 Unrelated Failures)"). Failures are specific to the fbcode sandcastle environment.

## Root Cause Analysis

### Issue 1: New test failure (0.074s — Python-level error)

The test uses `torch.utils.cpp_extension.load_inline` to compile and register a custom C++
extension at test time. In fbcode's sandcastle environment, this likely fails because:

- The sandcastle test runner doesn't have the JIT C++ compilation toolchain configured
the same way as OSS CI
- The buck target for `test_cpp_wrapper_custom_ops` may be missing proper dependencies
for runtime C++ compilation (`load_inline` requires headers + compiler)

The 0.074s failure time confirms this is a Python-level error (too fast for C++ compilation
of the cpp_wrapper output).

### Issue 2: Quantization test failures (the real bug)

The x86 quantized ops (e.g., `torch.ops.onednn.qconv2d_pointwise`) have schemas with
`int[]` (stride, padding, dilation), `str` (attr), and `Scalar?[]` (scalars) arguments.
These types are:

- **NOT** compatible with `_compatible_with_stableivalue` (it only supports base types + Optional)
- **ARE** compatible with `_compatible_with_cpp_boxed_dispatch` (supports ListType, StringType, etc.)

So these ops are now routed through the new boxed dispatch path instead of the Python fallback.

**The likely bug is in the output extraction from the stack after `callBoxed`:**

```python
# Current code in the PR:
for idx, output_arg in enumerate(output_args):
    if output_arg is None:
        continue
    dispatch_lines.writeline(
        f"{output_arg} = torch::aot_inductor::new_tensor_handle("
        f"std::move({stack_var}[{idx}]).toTensor());"
    )
```

After `callBoxed`, the stack contains ONLY the schema return values. But `output_args`
may include entries for `MutationOutput`s (in-place ops where the output is the same
tensor as an input). Using `enumerate(output_args)` to index the stack is incorrect when
there are MutationOutput entries — the stack index should only advance for non-mutation
outputs.

**Additionally**, there may be an IValue type issue: the PR constructs list arguments using
`c10::impl::GenericList`:

```python
dispatch_lines.writeline(
    f"c10::impl::GenericList {list_var}({type_expr(elem_type)});"
)
```

While modern PyTorch's `IValue::isIntList()` does check the `elementType` field (so
GenericList with IntType should work), some boxed kernel implementations may use typed
accessors that behave differently with GenericList vs specialized `c10::List<int64_t>`.

## Suggested Fix

```python
# Fix 1: Use a separate stack index counter for output extraction
stack_idx = 0
for output_arg, raw_output in zip(output_args, raw_outputs):
    if output_arg is None or isinstance(raw_output, ir.MutationOutput):
        continue
    dispatch_lines.writeline(
        f"{output_arg} = torch::aot_inductor::new_tensor_handle("
        f"std::move({stack_var}[{stack_idx}]).toTensor());"
    )
    stack_idx += 1
```

```python
# Fix 2: Use specialized c10::List types instead of GenericList for common element types
def codegen_list_ivalue(raw_arg, arg_type):
    elem_type = arg_type.getElementType()
    list_var = next_tmp("tmp_list")
    
    # Use specialized list types for better interop with boxed kernels
    if isinstance(elem_type, torch.IntType):
        dispatch_lines.writeline(f"c10::List<int64_t> {list_var};")
        dispatch_lines.writeline(f"{list_var}.reserve({len(raw_arg)});")
        for elem in raw_arg:
            dispatch_lines.writeline(
                f"{list_var}.push_back(static_cast<int64_t>("
                f"{self.val_to_arg_str_for_prim_type(elem, int)}));"
            )
    elif isinstance(elem_type, torch.FloatType):
        dispatch_lines.writeline(f"c10::List<double> {list_var};")
        dispatch_lines.writeline(f"{list_var}.reserve({len(raw_arg)});")
        for elem in raw_arg:
            dispatch_lines.writeline(
                f"{list_var}.push_back(static_cast<double>("
                f"{self.val_to_arg_str_for_prim_type(elem, float)}));"
            )
    elif isinstance(elem_type, torch.BoolType):
        dispatch_lines.writeline(f"c10::List<bool> {list_var};")
        dispatch_lines.writeline(f"{list_var}.reserve({len(raw_arg)});")
        for elem in raw_arg:
            dispatch_lines.writeline(
                f"{list_var}.push_back(static_cast<bool>("
                f"{self.val_to_arg_str_for_prim_type(elem, bool)}));"
            )
    elif isinstance(elem_type, torch.TensorType):
        dispatch_lines.writeline(f"c10::List<at::Tensor> {list_var};")
        dispatch_lines.writeline(f"{list_var}.reserve({len(raw_arg)});")
        for elem in raw_arg:
            tensor_expr = codegen_tensor_ivalue(elem, elem_type)
            # Extract tensor from IValue for direct list insertion
            dispatch_lines.writeline(
                f"{list_var}.push_back({tensor_expr}.toTensor());"
            )
    else:
        # Fallback to GenericList for complex/uncommon element types
        dispatch_lines.writeline(
            f"c10::impl::GenericList {list_var}({type_expr(elem_type)});"
        )
        dispatch_lines.writeline(f"{list_var}.reserve({len(raw_arg)});")
        for elem in raw_arg:
            dispatch_lines.writeline(
                f"{list_var}.emplace_back({codegen_ivalue(elem, elem_type)});"
            )
    
    return f"c10::IValue(std::move({list_var}))"
```

```python
# Fix 3: Make the new test skip in environments without JIT compilation support
@unittest.skipIf(not HAS_CPU, "requires CPU")
@unittest.skipIf(
    not torch.utils.cpp_extension._is_cuda_extension_build_available(),
    "requires C++ extension build support"
)
@config.patch(cpp_wrapper=True)
def test_cpp_custom_op_runtime_dispatch_stays_in_cpp(self):
    ...
```

Or better yet, restructure the test to use pre-registered ops from an existing library
rather than `load_inline`, which is fragile in buck-based build environments.

**Comment by jansel** [comment]
My agent says I addressed the internal reland analysis by fixing boxed dispatcher return indexing: mutation bookkeeping outputs are now skipped when mapping `callBoxed` stack returns, while real `None`/optional return slots still advance the stack index. I also skipped the inline `cpp_extension` regression in fbcode/sandcastle, where runtime C++ extension builds are unsupported, and expanded the OSS regression to cover mutation plus a separate boxed return and optional `None` return slots. Validation passed with `PYTHONPATH=. python test/inductor/test_cpp_wrapper_custom_ops.py TestCppWrapperCustomOps.test_cpp_custom_op_runtime_dispatch_stays_in_cpp`, targeted `lintrunner`, full `lintrunner -a`, and a fresh review subagent pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude read through the comments including the revert analysis.  Has the reason for the revert been addressed.  How confident are you this is safe to reland?
